### PR TITLE
NMS-16147: Extend confd support for Core container

### PR DESCRIPTION
**Pull request by mmahacek**
### All Contributors

* [X] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### External References

* Jira (Issue Tracker): https://opennms.atlassian.net/browse/NMS-16147

### CODE REVIEW

File: opennms-container/core/container-fs/confd/conf.d/kafka-broker.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[template]
+src = "kafka-broker.tpl"
+dest = "/usr/share/opennms/etc/opennms.properties.d/_kafka.properties"
```
Comment (mmahacek) [inline opennms-container/core/container-fs/confd/conf.d/kafka-broker.toml:3]:
```suggestion
dest = "/usr/share/opennms/etc/opennms.properties.d/_confd.kafka.properties"
```

File: opennms-container/core/container-fs/confd/conf.d/kafka-consumer-feature.toml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[template]
+src = "kafka-consumer-feature.tpl"
+dest = "/usr/share/opennms/etc/featuresBoot.d/kafka-consumer.boot"
```
Comment (mmahacek) [inline opennms-container/core/container-fs/confd/conf.d/kafka-consumer-feature.toml:3]:
```suggestion
dest = "/usr/share/opennms/etc/featuresBoot.d/_confd.kafka-consumer.boot"
```

**Review by ProjectJYL — APPROVED** [review]
Haven't tested it yet but it looks good to me.

**Review by garcia-sr — DISMISSED** [review]
LGTM

**Comment by mmahacek** [comment]
Moving back to draft - need to align Kafka settings with Minion Kafka config.

**Review by Bonrob2 — COMMENTED** [review]
Doc changes LGTM. Thanks!

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 3]
```diff
@@ -4,93 +4,432 @@
...
+| OPENNMS_DAEMONS_TELEMETRYD
+
+| OPENNMS_DAEMONS_TICKETER
+
+| OPENNMS_DAEMONS_TRAPD
 |===
+
+=== RRD/RRA definitions
+
+Data collection retention RRA definitions can only be defined via `{page-component-name}-config.yaml` and not by environment variables.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:330]:
```suggestion
You can define data collection retention RRA definitions only via `{page-component-name}-config.yaml` and not by environment variables.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 4]
```diff
@@ -4,93 +4,432 @@
...
+If RRA definitions are provided, any `<rra>` elements will be replaced across all `\*datacollection*` files and any files located in `$\{OPENNMS_HOME}/etc/\*datacollection*.d/` subdirectories.
+This does not change the RRA definitions for pollerd services.
+
+IMPORTANT: If you change the RRA definitions once datacollection has already run, you will need to manually delete any existing RRD/JRB files so they can be recreated with the new definition.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:334]:
```suggestion
IMPORTANT: If you change the RRA definitions after datacollection has already run, you will need to manually delete any existing RRD/JRB files so they can be recreated with the new definition.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 5]
```diff
@@ -4,93 +4,432 @@
...
+No new metrics will be collected until the files are deleted or the RRA definitions reverted back to their previous values.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:335]:
```suggestion
No new metrics will be collected until the files are deleted or the RRA definitions revert back to their previous values.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 6]
```diff
@@ -4,93 +4,432 @@
...
+      - RRA:AVERAGE:0.5:1:2016
+      - RRA:AVERAGE:0.5:12:1488
+      - RRA:AVERAGE:0.5:288:366
+      - RRA:MAX:0.5:288:366
+      - RRA:MIN:0.5:288:366
+----
+
+=== Kafka broker configuration
+
+Kafka broker settings can only be defined via `{page-component-name}-config.yaml` and not by environment variables.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:353]:
```suggestion
You can define Kafka broker settings only via `{page-component-name}-config.yaml` and not by environment variables.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 7]
```diff
@@ -4,93 +4,432 @@
...
+<1> If modifying the instance ID, make sure to set the same ID on the Minion configs.
+<2> If you have multiple Kafka brokers in your cluster, {page-component-title} will attempt to connect to each listed server in order.
+The first broker server to respond will be used to provide a full list of all brokers in the cluster.
+<3> Any settings defined within this section will be written as `org.opennms.core.ipc.kafka.key=value`.
+<4> You can also define `ipc.rpc.kafka.\*`, `ipc.sink.kafka.\*`, and `ipc.twin.kafka.\*`, with a similiar structure, if you want to define different settings for each of those systems.
+<5> Any settings defined within this section will be written as `org.opennms.core.ipc.sink.kafka.key=value`.
+
+=== Kafka consumer configuration
+
+Kafka Consumer settings can only be defined via `{page-component-name}-config.yaml` and not by environment variables.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:380]:
```suggestion
You can define Kafka consumer settings only via `{page-component-name}-config.yaml` and not by environment variables.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 8]
```diff
@@ -4,93 +4,432 @@
...
+  kafka:
+    consumer:
+      topic: opennms-kafka-events <1>
+      bootstrap.servers: <2>
+        - broker1:9092
+        - broker2:9092
+      key: value <3>
+----
+<1> Topic will be written to `$\{OPENNMS_HOME}/etc/org.opennms.features.kafka.consumer.cfg`.
+<2> If you have multiple Kafka brokers in your cluster, Kafka Consumer will attempt to connect to each listed server in order.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:394]:
```suggestion
<2> If you have multiple Kafka brokers in your cluster, Kafka consumer will attempt to connect to each listed server in order.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 9]
```diff
@@ -4,93 +4,432 @@
...
+The first broker server to respond will be used to provide a full list of all brokers in the cluster.
+This setting will be written to `$\{OPENNMS_HOME}/etc/org.opennms.features.kafka.consumer.client.cfg`.
+`$\{OPENNMS_HOME}/etc/featuresBoot.d/kafka-consumer.boot` will also be created when bootstrap servers have been defined.
+<3> Any settings defined within this section will be written as `key=value` to `$\{OPENNMS_HOME}/etc/org.opennms.features.kafka.consumer.client.cfg`.
+
+=== Elastic flows configuration
+
+Elastic settings can only be defined via `{page-component-name}-config.yaml` and not by environment variables.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:402]:
```suggestion
You can define Elastic settings only via `{page-component-name}-config.yaml` and not by environment variables.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 10]
```diff
@@ -4,93 +4,432 @@
...
+<1> If an instance ID is present, it will be used as an index prefix.
+<2> If you have multiple Elastic servers in your cluster, {page-component-title} will attempt to connect to each listed server in order.
+The first  server to respond will be used to provide a full list of all servers in the cluster.
+<3> If your Elastic servers require credentials, you can define `username` and `password` values which will be written to `$\{OPENNMS_HOME}/etc/elastic-credentials.xml`.
+Alternatively, you can define `globalElasticUser` and `globalElasticPassword` in the config section to define one set of credentials to use across all Elastic servers.
+<4> Any settings defined within this section will be written as `key=value` to `$\{OPENNMS_HOME}/etc/org.opennms.features.flows.persistence.elastic.cfg`.
+
+=== System properties
+
+System properties can only be defined via `{page-component-name}-config.yaml` and not by environment variables.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:426]:
```suggestion
You can define system properties only via `{page-component-name}-config.yaml` and not by environment variables.
```

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 11]
```diff
@@ -4,93 +4,432 @@
...
+=== Daemon control
+
+Select daemons are able to be enabled or disabled through the container definition with the following boolean variables.
 
 [options="header, %autowidth"]
 |===
-| Mountpoint             | Description
-| `/opt/opennms-overlay` | Allows you to overwrite files relative to `/opt/opennms`.
-| `/opennms-data`        | Directory with RRDTool/JRobin files and generated PDF reports sent to the file system.
+| Environment variable
```
Comment (Bonrob2) [inline docs/modules/reference/pages/configuration/core-docker.adoc:309]: This looks odd as a table with only one cell. Maybe make it a bulleted list?

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
A few updates. Thanks!

File: docs/modules/reference/pages/configuration/core-docker.adoc [hunk 12]
```diff
@@ -4,93 +4,432 @@
...
+| OPENNMS_DAEMONS_SNMPPOLLER
+
+| OPENNMS_DAEMONS_SYSLOGD
+
+| OPENNMS_DAEMONS_TELEMETRYD
+
+| OPENNMS_DAEMONS_TICKETER
+
+| OPENNMS_DAEMONS_TRAPD
 |===
```
Comment (mmahacek) [inline docs/modules/reference/pages/configuration/core-docker.adoc:326]:
```suggestion
* OPENNMS_DAEMONS_CORRELATOR
* OPENNMS_DAEMONS_ENLINKD
* OPENNMS_DAEMONS_SNMPPOLLER
* OPENNMS_DAEMONS_SYSLOGD
* OPENNMS_DAEMONS_TELEMETRYD
* OPENNMS_DAEMONS_TICKETER
* OPENNMS_DAEMONS_TRAPD
```

File: docs/modules/reference/pages/configuration/core-docker.adoc (hunk 11 above)
Comment (mmahacek) [reply docs/modules/reference/pages/configuration/core-docker.adoc:309]: That can be done.  I had made it a table to fit with the format of the other sections, but a list does make more sense here.

**Review by Bonrob2 — APPROVED** [review]
docs LGTM. Thanks!

**Review by indigo423 — COMMENTED** [review]
Just some food for thought, here. I would be careful for configs which can be changed not just through the file system, such as all Karaf plugin configs, .e.g. `org.opennms.features.flows.persistence.elastic.cfg`. They can be modified through the Karaf shell and can lead to some problems. You need to tell users to not edit these configs via Karaf. Establishing also awareness for our developers to not touch configs in Karaf plugins without updating the templates is also pretty challenging.

**Comment by mmahacek** [comment]
If only these cfg files were pulled into the database to make them more persistent between container/pod rebuilds....

**Pull request closed without merging**